aten/src/ATen/core/jit_type.h

```diff
@@ -846,11 +847,34 @@ struct CAFFE2_API FutureType : public SingleElementType<TypeKind::FutureType, Fu
 
 struct TupleType;
 using TupleTypePtr = std::shared_ptr<TupleType>;
-using OptNameList = c10::optional<std::vector<std::string>>;
+using NameList = std::vector<std::string>;
 // This type represents a Tuple
 struct CAFFE2_API TupleType : public Type {
-  static TupleTypePtr create(std::vector<TypePtr> types, OptNameList names=c10::nullopt, c10::optional<std::string> unqualName=c10::nullopt) {
-    return TupleTypePtr(new TupleType(std::move(types), std::move(names), std::move(unqualName))); // NOLINT(modernize-make-shared)
+  struct CAFFE2_API NamedTupleSpec {
+    // Storing types here too is redundant but it makes encapsulating NamedTuple functionality
+    // easier.
+    NamedTupleSpec(std::vector<TypePtr> types, NameList names, c10::optional<c10::QualifiedName> qualName=c10::nullopt)
+      : types(std::move(types)), names(std::move(names)), qualName(std::move(qualName)) {
+      createFunctionSchema();
+    }
+
+    std::vector<TypePtr> types;
```

Comment:
This class needs to be encapsulated. Public members will not age well. Also it seems to have redundant info for instance a FunctionSchema and a list of types and names. That is exactly what an Argument list stores. I'd prefer we just kept a FunctionSchema.